posts/inside-rust/2023-07-11-trait-system-refactor-initiative.md

```diff
@@ -0,0 +1,47 @@
+
+It has now been half a year since that announcement which matches the first step of [our roadmap][roadmap].
+
+## The status quo
+
+The new trait solver implementation can be tested on nightly by using the rustc flag `-Ztrait-solver=next`. To only use the new implementation for coherence, `-Ztrait-solver=next-coherence` can be used. See the current implementation progress of the new trait solver in [its tracking issue](https://github.com/rust-lang/rust/issues/107374).
+
+We are now at a point where the we exclusively rely on the new implementation when the feature flag is enabled. This is a major step as we've previously still relied on the old solver for ["deep normalization"](https://github.com/rust-lang/rust/pull/113086) and ["selection"](https://github.com/rust-lang/rust/pull/112869). When using the new trait solver many crates and most of our existing regression tests successfully compile.
+
+While there is a significant tail of less common bugs, we currently have two main failure causes. The new solver has a slightly different approach for `impl Trait`, the implementation of which is still broken for nested RPIT, e.g. `fn foo() -> impl Trait<Assoc = impl Sized>`. Implementing this new approach helped us to discover issues of the existing implementation which allows us to refine its design before the stabilization of TAIT. The inference of implicit unsize coercions, e.g. converting `Box<String>` to `Box<dyn Display>`, relies on implementation details of the existing trait solver. We've recently started to emulating the existing behavior here and should fix most of the remaining breakage from that over the next few weeks.
+
+## Going forward
+
+After fixing the currently open issues, we intend to move parts of rustc to the new trait solver implementation in steps, starting by using it in coherence. We expect to move coherence to the new implementation at the end of this year. Moving the type checking of functions to the new trait solver implementation will be a lot more challenging than coherence. This will be the last place where we will use the old implementation. We expect to change the default there in 2024, potentially relying on the new edition to help with migration.
+
+A major challenge will be "incompleteness". The trait system sometimes unnecessarily guides type inference. Consider the following example:
```

Comment:
```suggestion
A major challenge will be "incompleteness" (note the usage of "incompleteness" here is not the colloquial definition but instead a term of art with a precise technical meaning). The trait system sometimes unnecessarily guides type inference. Consider the following example:
```

Reply:
This leaves me wondering what the technical meaning of "incompleteness" is. Is there a simplified summary that could be included for those of us who are curious?

```suggestion
A major challenge will be "incompleteness". In this context, incompleteness is a technical term which means <short explanation>. The trait system sometimes unnecessarily guides type inference. Consider the following example:
```

Reply:
updated, does this explanation look useful?

Reply:
Yes, thank you!